design-documents/distributed-deployment.md

```diff
@@ -0,0 +1,217 @@
+
+In both cases database upgrade need to be run only on admin instance or CLI instance.
+
+### Decomposing monolith
+
+Area specific instance will contain only relevant set of modules. For instance, admin instance will contain *AdminUi modules and their dependencies and will not contain store front modules.
+
+Currently Magento 2 project depends on `magento/base` component, that contain project structure and entry points. We need to decompose it in the following way to not have unnecessary entry points on each instance.
+`magento/base` (will not contain area specific code anymore)
+`magento/web-application` - will contain index.php endpoint, for admin, store front and web API (if web API going to use the same endpoint in the future)
+`magento/cron-application` - will contain cron.php endpoint
+`magento/webapi-application` - will contain endpoint for webapi, if we decide to have separate
+
+![Base package decomposition](distributed-deployment/base-package-decomposition.png)
+
+Installing all components will result in monolith, project for monolith will depend on all of the components. Installing `magento/base` and `magento/web-application` and Admin UI modules will result in admin instance.
+
+In addition to base package decomposition, we need to add marker components that would identify area specific instance:
+`magento/cron` - marker component, contains cron modules
+`magento/admin-ui` - marker component, contains admin ui modules
+`magento/ui` - marker component, contains store front modules
+
+Different options for area specific instances deployment
+
+1. Distribute different projects and use composer to manage dependencies
+    
+    UI modules will have to depend on marker component area, so it would be possible to identify on which area specific instance they need to be installed. Modules that are reusable across instances will not have dependency on area specific instances. UI modules will depend on them and they will be included where they needed.
+    
+    There are can be a case when extension is not necessarily depends on UI module but needs to be installed on particular instance or instances (for instance, extension that sends an email when the order is placed), in this place it will depend on base module (module that contains business logic) and will be installed on all instances.
+    
+    Pros
+    * Simplifies developer workflow (see below)
+    
+    Cons
+    * Will require us to publish more projects
+    * Potentially some modules that don't depend on UI and depend on base modules may be installed on all instances (need to investigate)
+    * May not work for distributed deployment of services (or one to many relationship, when module need to be installed on multiple instances), see diagram below
+    * How to manage obsolete dependencies (you remove module on once instance, need to remove relevant parts on another)?
```

Comment:
As for distributed deployment, we can make only sync upgrade, I think it should not be a problem for the removing packages as it can be done by the composer on each instance.

Reply:
Wouldn't we need to track what packages need to be removed on each instance and remove them manually?

Reply:
Actually, we need. As we will upgrade all dependencies synchronously, probably, removing we will do in the same way.